llvm/unittests/ADT/FunctionExtrasTest.cpp

```diff
@@ -310,4 +310,25 @@ class Incomplete {};
 Incomplete incompleteFunction() { return {}; }
 const Incomplete incompleteFunctionConst() { return {}; }
 
+// Check that the moved-from captured state is properly destroyed during
+// move construction/assignment.
+TEST(UniqueFunctionTest, MovedFromStateIsDestroyedCorrectly) {
+  static int NumOfMovesCalled = 0;
+  static int NumOfDestructorsCalled = 0;
+  struct State {
+    State() = default;
+    State(const State &) = delete;
+    State(State &&) { ++NumOfMovesCalled; }
+    State &operator=(const State &) = delete;
+    State &operator=(State &&Rhs) { return *new (this) State{std::move(Rhs)}; }
+    ~State() { ++NumOfDestructorsCalled; }
+  };
+  {
+    unique_function<void()> CapturingFunction{[state = State{}] {}};
+    unique_function<void()> CapturingFunctionMoved{
+        std::move(CapturingFunction)};
+  }
+  EXPECT_EQ(NumOfDestructorsCalled, 1 + NumOfMovesCalled);
```

Comment:
Might be ore obvious/clear to have the default ctor count too - maybe count all the ctor calls together, then check that the dtor calls == ctor calls. 

But also, might be worth checking - I think we have a reusable special member counting type utility used in other tests that could be reused here? llvm/unittests/ADT/CountCopyAndMove.h - though, admittedly, we haven't been great about using that everywhere, and various data structure tests certainly have their own versions, but maybe we can try to stem the tide of more of them?

Reply:
Sounds good to me, refactored the test to use that helper. Had to upgrade it a bit to include default constructions as well.

While testing I think I might have actually found another bug (which can also explain why this destructor issue didn't come up before). There is this check in the code:

```
if (sizeof(CallableT) > InlineStorageSize ||
    alignof(CallableT) > alignof(decltype(StorageUnion.InlineStorage))) {
    IsInlineStorage = false;
    // ...
}
```

Where `InlineStorage` is defined as:

```
alignas(void *) mutable std::byte InlineStorage[InlineStorageSize];
```

I believe this expression `alignof(decltype(StorageUnion.InlineStorage))` actually returns 1 and so we always skip the inline storage code-path (CompilerExplorer [seems to agree](https://godbolt.org/z/oEPsfTqWK)).

I've fixed that by using an explicit `InlineStorageAlign` variable but let me know if you want me to split this PR into two.

Reply:
Testing looks great - thanks for that!

Yeah, could you split out the alignment problem and we'll discuss that separately - I'm not quite following it/would like to understand it better.

Reply:
Sure thing, removed the alignment-related code from this PR. Opened a separate one here - https://github.com/llvm/llvm-project/pull/99403